File: leet-code/111.二叉树的最小深度.js
/*
 * @lc app=leetcode.cn id=111 lang=javascript
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
  if (!root) return 0
  const leftHight = minDepth(root.left)
  const rightHight = minDepth(root.right)
  if (!root.left && root.right) {
    return rightHight + 1
  }
  if (!root.right && root.left) {
    return leftHight + 1
  }
  return Math.min(leftHight, rightHight) + 1
};
// @lc code=end

File: leet-code-rs/src/solutions/111_二叉树的最小深度.rs
/*
 * @lc app=leetcode.cn id=111 lang=rust
 *
 * [111] 二叉树的最小深度
 */
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

#[allow(unused)]
impl Solution {
    pub fn min_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        match root {
            Some(node) => {
                let node = node.borrow();
                if node.left.is_none() {
                    return Self::min_depth(node.right.clone()) + 1;
                }

                if node.right.is_none() {
                    return Self::min_depth(node.left.clone()) + 1;
                }

                let l_depth = Self::min_depth(node.left.clone());
                let r_depth = Self::min_depth(node.right.clone());

                return l_depth.min(r_depth) + 1;
            }
            None => 0,
        }
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_min_depth() {
        //assert_eq!(Solution::min_depth(vec![]),[]);
        assert!(true)
    }
}
